docs/developers/napari_architecture.md

```diff
@@ -0,0 +1,200 @@
+attributes are changed an "event" needs to be emitted such that relevant
+obsevers of the model (such as other classes) can take the appropriate
+action.
+
+One way this is achieved in napari is via getters and setters. Let's take
+for example the `Dims` class with a selected few attributes:
+
+```python
+from napari.utils.events import EmitterGroup
+
+class Dims:
+    """Dimensions object modeling slicing and displaying.
+
+    Parameters
+    ----------
+    ndim : int
+        Number of dimensions.
+    ndisplay : int
+        Number of displayed dimensions.
+    ...
+    """
+    def __init__(self, ndim, ndisplay):
+        self._ndim_ = ndim
+        self._ndisplay = ndisplay
+
+        # an `EmitterGroup` manages a set of `EventEmitters`
+        # we add one emitter for each attribute we'd like to track
+        self.events = EmitterGroup(source=self, ndim=None, ndisplay=None)
+
+    # for each attribute, we create a `@property` getter/setter
+    # so that we can emit the appropriate event when that attribute
+    # is changed using the syntax: ``Dim.attribute = new_value``
+    @property
+    def ndim(self):
+        """Number of dimensions."""
+        return self._ndim
+
+    @ndim.setter
+    def ndim(self, value):
+        self._ndim = value
+        # emit the ndim "changed" event
+        self.events.ndim(value=value)
+
+    @property
+    def ndisplay(self):
+        """Number of displayed dimensions."""
+        return self._ndisplay
+
+    @ndisplay.setter
+    def ndisplay(self, value):
+        self._ndisplay = value
+        # emit the ndisplay "changed" event
+        self.events.ndisplay(value=value)
+```
+
+Another object can then "listen" for changes in our `Dim` model and register
+a callback function with the event emitter of the attribute they would like
+to watch:
+
+```python
+# create an instance of the model
+dims = Dim(ndim=3, ndisplay=2)
+
+# define some callback that should respond to changes in the model
+def _update_display(self):
+    """
+    Updates display for all sliders.
+
+    The event parameter is there just to allow easy connection to signals,
+    without using `lambda event:`
+    """
+    # the code updating display code is quite complex and not relevant for this
+    # example thus has ommited.
+    nsteps = self.dims.nsteps
+    print(f"Update number of dimensions displayed to {nsteps}")
+
+# register our callback with the model
+dims.events.wind.connect(_update_display)
+
+# now, everytime dims.ndisplay is changed, _update_display is called
+dim.ndisplay = 3
+```
+
+This method is very customizable but requires a lot of boilerplate. The
+generic base model `EventedModel` was added to reduce this and
+"standardize" this change/emit pattern. The `EventedModel` provides the
+following features:
+
+* type validation and coersion on class instantiation and attribute assignment
+* event emission after successful attribute assignment
+
+Using `EventedModel` would reduce the above `Dim` class code to:
```

Comment:
In the previous example, the callback made it clear what was happening when something is changed. Could you clarify how this works with `EventedModel`? Do you still need to connect the callback with this new Dim definition?

Reply:
Ah yes good point!